dask/config.py

```diff
@@ -0,0 +1,194 @@
+from __future__ import print_function, division, absolute_import
+
+from contextlib import contextmanager
+import os
+import sys
+
+from .compatibility import FileExistsError
+
+
+config_paths = [
+    '/etc/dask',
+    os.path.join(sys.prefix, 'etc', 'dask'),
+    os.path.join(os.path.expanduser('~'), '.config', 'dask'),
+    os.path.join(os.path.expanduser('~'), '.dask')
+]
+
+if 'DASK_CONFIG' in os.environ:
+    config_paths.append(os.environ['DASK_CONFIG'])
+
+
+def update(old, new):
+    """ Update a nested dictionary with values from another
+
+    This is like dict.update except that it smoothly merges nested values
+
+    This operates in-place and modifies old
+
+    Example
+    -------
+    >>> a = {'x': 1, 'y': {'a': 2}}
+    >>> b = {'y': {'b': 3}}
+    >>> update(a, b)
+    {'x': 1, 'y': {'a': 2, 'b': 3}}
+
+    See Also
+    --------
+    merge
+    """
+    for k, v in new.items():
+        if k not in old and type(v) is dict:
+            old[k] = {}
+
+        if type(v) is dict:
+            update(old[k], v)
+        else:
+            old[k] = v
+    return old
+
+
+def merge(*dicts):
+    """ Update a sequence of nested dictionaries
+
+    This prefers the values in the latter dictionaries to those in the former
+
+    Example
+    -------
+    >>> a = {'x': 1, 'y': {'a': 2}}
+    >>> b = {'y': {'b': 3}}
+    >>> merge(a, b)
+    {'x': 1, 'y': {'a': 2, 'b': 3}}
+
+    See Also
+    --------
+    update
+    """
+    result = {}
+    for d in dicts:
+        update(result, d)
+    return result
+
+
+def collect_yaml(paths=config_paths):
+    """ Collect configuration from yaml files
+
+    This searches through ``config_paths``, expands to find all yaml or json
+    files, and then parses each file.
+    """
+    # Find all paths
+    file_paths = []
+    for path in paths:
+        if os.path.exists(path):
+            if os.path.isdir(path):
+                file_paths.extend(sorted([
+                    p for p in os.listdir(path)
+                    if os.splitext(p)[1].lower() in ('json', 'yaml', 'yml')
+                ]))
+            else:
+                file_paths.append(path)
+
+    configs = []
+
+    # Parse yaml files
+    for path in file_paths:
+        with open(path) as f:
+            data = yaml.load(f.read() or {})
```

Comment:
`yaml.load(f.read()) or {}` - but probably should have try/except (with logger warnings), so syntax errors in the files don't break config.

Reply:
Indeed.  Thanks for the `{}` fixs .  

I'm inclined to fail hard on syntax errors in yaml files.  This is something that users opt-in to, so I think it's ok to give them immediate hard feedback.